redirector.rs/src/main.rs

```diff
@@ -0,0 +1,180 @@
+            let target = mountpoint.join(subdir);
+            reply.data(target.as_os_str().as_bytes());
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+
+    fn readdir(
+        &mut self,
+        _req: &Request,
+        ino: u64,
+        _fh: u64,
+        offset: i64,
+        mut reply: ReplyDirectory,
+    ) {
+        if ino == ROOT_DIR_INODE {
+            if offset == 0 {
+                reply.add(ROOT_DIR_INODE, 0, FileType::Directory, ".");
+                reply.add(ROOT_DIR_INODE, 1, FileType::Directory, "..");
+                reply.add(PRIVATE_LINK_INODE, 2, FileType::Symlink, "private");
+                reply.add(PUBLIC_LINK_INODE, 3, FileType::Symlink, "public");
+                reply.add(TEAM_LINK_INODE, 4, FileType::Symlink, "team");
+            }
+            reply.ok();
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+}
+
+// This is explicitly only implemented for Linux, so that building on other
+// platforms is a compiler error.
+#[cfg(target_os = "linux")]
+fn get_keybase_fuse_mount(uid: u32) -> Option<PathBuf> {
+    let uid_str = format!("user_id={}", uid);
+    let file = File::open("/proc/mounts").expect("failed to open /proc/mounts");
+    let reader = BufReader::new(file);
+    // TODO: Avoid assuming UTF8?
+    // TODO: Sort the paths first?
+    for line in reader.lines() {
+        let line = line.expect("failed to read from /proc/mounts");
+        let parts: Vec<&str> = line.split_whitespace().collect();
+        // Check the mount type.
+        if parts[2] != "fuse" {
+            continue;
+        }
+        // Look for the uid in the mount options.
+        if parts[3].split(',').find(|s| s == &uid_str).is_none() {
+            continue;
+        }
+        // Make sure "keybase" is in the path somewhere. We do this loosely,
+        // rather than splitting exactly on slashes, to handle "keybase.devel"
+        // etc.
+        if parts[1].to_lowercase().find("keybase").is_none() {
+            continue;
+        }
+        let path: PathBuf = parts[1].into();
+        // TODO: We want to check that .kbfs_error exists, but the kbfsfuse
+        // mount doesn't currently set "allow_root".
+        return Some(path);
+    }
+    None
+}
+
+fn main() {
+    // The rust-fuse library writes logs using the standard log crate
+    // interface. To init a log backend that will actually print those,
+    // uncomment this line:
+    // env_logger::Builder::new().filter(None, log::LevelFilter::Debug).init();
+
+    let redirector = Redirector {
+        start_time: time::get_time(),
+    };
+    let mountpoint = env::args_os().nth(1).expect("must have a path argument");
+    let options = &["-o", "allow_other", "-o", "fsname=keybase-redirector"];
+    let options_osstr: Vec<&OsStr> = options.iter().map(|o| o.as_ref()).collect();
+    fuse::mount(redirector, &mountpoint, &options_osstr).expect("mount returned an error");
```

Comment:
I might be misunderstanding something important. Doesn't the SETUID bit in the file permissions take care of this?

Reply:
Nope!  The program still has to do an explicit setuid call when it starts up: https://doc.rust-lang.org/1.4.0/libc/funcs/posix88/unistd/fn.setuid.html

You can try it out by just running the binary without `sudo` and seeing if it can mount (it can't).